allennlp/models/encoder_decoders/wikitables_semantic_parser.py

```diff
@@ -145,25 +152,88 @@ def forward(self,  # type: ignore
+        batch_size = table_text['tokens'].size(0)
+        max_num_entities = table_text['tokens'].size(1)
+        num_types = 2  # todo should we compute from type_signatures?
+
         # (batch_size, question_length, embedding_dim)
-        embedded_input = self._question_embedder(question)
+        embedded_question = self._question_embedder(question)
         question_mask = util.get_text_field_mask(question).float()
-        batch_size = embedded_input.size(0)
 
-        # Actually a Dict[str, torch.LongTensor], but there is probably a single entry, with a
-        # tensor of shape (batch_size, num_entities, num_entity_tokens).
-        table_text = table['text']
-        embedded_table_text = self._question_embedder(table_text)
-        # TODO(mattg): embed the table, similar to how the question is embedded (probably just
-        # using the question embedder), giving a tensor of shape (batch_size, num_entities,
-        # num_entity_tokens, embedding_dim).  Then encode that into shape (batch_size,
-        # num_entities, embedding_dim).
+        # (batch_size, num_entities, num_entity_tokens, embedding_dim)
+        embedded_table = self._question_embedder(table_text)
+        # (batch_size, num_entities, num_entity_tokens)
+        table_mask = util.get_text_field_mask(table_text, num_wrapping_dims=1).float()
+
+        # compute the average of each entity's word embeddings
+        # (batch_size, num_entities, embedding_dim)
+        encoded_table = self._compute_average_with_encoder(embedded_table, table_mask)
+
+        # (batch_size, num_entities, num_entities)
+        neighbor_idx_tensor = self._get_neighbor_indexes(world, max_num_entities)
+
+        # get averaged word embeddings for each entity's neighbors
+        # (batch_size, num_entities, num_entities, embedding_dim)
+        neighbor_embed_tensor = util.batched_index_select(encoded_table, neighbor_idx_tensor)
+        neighbor_mask = util.get_text_field_mask({'tokens': neighbor_idx_tensor},
+                                                 num_wrapping_dims=1).float()
+
+        # get averaged neighbor embeddings for each entity
+        # (batch_size, num_entities, embedding_dim)
+        neighbor_embeddings = self._compute_average_with_encoder(neighbor_embed_tensor, neighbor_mask)
+
+        # each entity has a binary flag indicating type
+        # (batch_size, num_entities)
+        type_vector = self._get_type_vector(world, max_num_entities)
+
+        # construct the entity embeddings through a linear combination of type
+        # and neighbor embeddings fed through tanh
+        type_params = torch.nn.Linear(num_types, embedding_dim)
+        neighbor_params = torch.nn.Linear(embedding_dim, embedding_dim)
+        x = type_params(type_vector.float())
+        y = neighbor_params(neighbor_embeddings.float())
+        tan = torch.nn.Tanh()
+        # (batch_size, num_entities, embedding_dim)
+        entity_embeddings = tan(x + y)
```

Comment:
You can just do [`entity_embeddings = torch.nn.functional.tanh(x + y)`](http://pytorch.org/docs/master/nn.html#torch.nn.functional.tanh), instead of instantiating a `Tanh` object.